src/data/adding_prob_src.rs

```diff
@@ -0,0 +1,134 @@
+extern crate rand;
+use rand::distributions::{Range, IndependentSample};
+use af;
+use af::{Array, Dim4, MatProp, DType};
+use std::cell::{RefCell, Cell};
+
+use initializations::uniform_pos;
+use utils;
+
+use data::{Data, DataSource, DataParams, Normalize, Shuffle};
+
+pub struct AddingProblemSource {
+    pub params: DataParams,
+    pub iter: Cell<u64>,
+    pub offset: Cell<f32>,
+    pub bptt_unroll : u64, 
+}
+
+impl AddingProblemSource {
+    pub fn new(batch_size: u64, bptt_unroll: u64, dtype: DType, max_samples: u64) -> AddingProblemSource
+    {
+        let input_dims = Dim4::new(&[batch_size, 1, bptt_unroll, 1]);
+        let target_dims = Dim4::new(&[batch_size, 1, 1, 1]);
+        let train_samples = 0.7 * max_samples as f32;
+        let test_samples = 0.2 * max_samples as f32;
+        let validation_samples = 0.1 * max_samples as f32;
+        AddingProblemSource {
+            params : DataParams {
+                input_dims: input_dims,
+                target_dims: target_dims,
+                dtype: dtype,
+                normalize: false,
+                shuffle: false,
+                current_epoch: Cell::new(0),
+                num_samples: max_samples,
+                num_train: train_samples as u64,
+                num_test: test_samples as u64,
+                num_validation: Some(validation_samples as u64),
+            },
+            iter: Cell::new(0),
+            offset: Cell::new(0.0f32),
+            bptt_unroll: bptt_unroll,
+        }
+    }
+
+    fn generate_input(&self, batch_size: u64, bptt_unroll: u64) -> Array {
+        let dim1 = Dim4::new(&[batch_size, 1, bptt_unroll/2, 1]);
+        let ar1 = uniform_pos::<f32>(dim1, 1.0);
+
+        let between1 = Range::new(0, bptt_unroll/4);
+        let between2 = Range::new(bptt_unroll/4, bptt_unroll/2);
+        let mut rng1 = rand::thread_rng();
+        let mut rng2 = rand::thread_rng();
+
+        let mut vec_total = Vec::new();
+        let vec = vec!(0f32; (bptt_unroll/2) as usize);
+
+        for _ in 0..batch_size {
+            let index1 = between1.ind_sample(&mut rng1) as usize;
+            let index2 = between2.ind_sample(&mut rng2) as usize;
```

Comment:
This can be cleanly represented by an inner loop over 2 elements. Never repeat code.